### DISCUSSION TITLE
How to enforce a single line

### DISCUSSION THREAD

**Discussion by vnugent** [facebook/lexical#3640] [Q&A] [answered]
How to create an editor that behaves like a HTML input field, preventing users from hitting enter to create multiple lines?

**Comment by moy2010** [comment, marked as answer]
Check the [quick-start docs](https://lexical.dev/docs/getting-started/quick-start). You can create the editor as in the vanilla-JS example and set an HTML input field as the root element.

**Comment by vnugent** [comment]
I ended up writing a plugin (inspired by the playground [MaxLengthPlugin](https://github.com/facebook/lexical/tree/main/packages/lexical-playground/src/plugins/MaxLengthPlugin)).

It might not as efficient as the way the MaxLength plugin caches and restores the previous state when a condition is met.  But it seems to serve me well.


```javascript
import { useLexicalComposerContext } from '@lexical/react/LexicalComposerContext'
import { $createParagraphNode, $createTextNode, RootNode } from 'lexical'
import { useEffect } from 'react'

const newlinesRegex = /[\n\r]/g

/**
 * Enforce a single line of text by removing new line characters (from enter key or copy-n-paste content)
 */
export function SingleLinePlugin (): null {
  const [editor] = useLexicalComposerContext()

  useEffect(() => {
    return editor.registerNodeTransform(RootNode, (rootNode: RootNode) => {
      const textContent = rootNode.getTextContent()

      // test if content has new line character(s)
      if (newlinesRegex.test(textContent)) {
        // yes, remove all new line chars
        const newText = textContent.replace(newlinesRegex, '')

        // replace current content
        const paragraph = $createParagraphNode()
        paragraph.append(
          $createTextNode(newText)
        )
        rootNode.clear().append(paragraph)
        rootNode.selectEnd() // move cursor to end of text
      }
    })
  }, [editor])

  return null
}
```

**Reply by shota-6** [reply]
@vnugent 
Thanks for sharing the useful plugin!
I have a question.
If I press enter while the text has a style, such as size, color, etc., all of them are canceled. Is there a solution for this?

**Reply by nicolasthy** [reply]
Managed to make it work with this:
```js
import { useEffect } from "react";

import { useLexicalComposerContext } from "@lexical/react/LexicalComposerContext";
import { $createParagraphNode, RootNode } from "lexical";

const newlinesRegex = /[\n\r]/g;

export function SingleLinePlugin(): null {
  const [editor] = useLexicalComposerContext();

  useEffect(() => {
    return editor.registerNodeTransform(RootNode, (rootNode: RootNode) => {
      const nodes = rootNode.getAllTextNodes();
      const textContent = rootNode.getTextContent();

      if (newlinesRegex.test(textContent)) {
        const paragraph = $createParagraphNode();
        paragraph.append(...nodes);
        rootNode.clear().append(paragraph);
        rootNode.selectEnd();
      }
    });
  }, [editor]);

  return null;
}
```

Instead I'm only checking new lines from the textContent and then if there's a new line I rewrite the content using the nodes (they don't come with new lines).

However I'm unable to have the input stay on a single line, it automatically wraps...
<img width="577" alt="image" src="https://github.com/facebook/lexical/assets/4039090/f128ba21-b716-4e83-9a95-7d13a4a962c6">

If anyone finds a solution for this, that would be great ! 🙌

**Reply by moy2010** [reply]
@nicolasthy . The visual wrapping can be fixed with CSS like this:

```
white-space: nowrap;
```

**Reply by nicolasthy** [reply]
That I know, but Lexical's `<ContentEditable>` already adds it to the styles, so we can't easily override the value:
```html
<div 
  class="prose prose-sm overflow-y-auto px-3 py-2 focus:outline-none" 
  contenteditable="true" 
  spellcheck="true" 
  data-lexical-editor="true" 
  style="user-select: text; white-space: pre-wrap; word-break: break-word;" 
  role="textbox">
</div>
```

**Reply by moy2010** [reply]
Got it. I would look into the theming part of Lexical for any CSS override needs. Hard to say if it would work off the top of my head tbh.

**Reply by nicolasthy** [reply]
Found out it was set in `setRootElement` in LexicalEditor.ts, so guess I'll have a hard time fixing this 😞
